aarch64/kernel/memory_aarch64/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+            // or to use them in the page table walk ("Used").
+            // With our four-level paging, however, the top-byte
+            // is not used for page table walks anyway.
+              TCR_EL1::TBI0::Used
+            // | TCR_EL1::TBI1::Used
+
+            // Translation Granule Size = Page Size
+            // => four kilobytes
+            + TCR_EL1::TG0::KiB_4
+            // + TCR_EL1::TG1::KiB_4
+
+            // These fields could only be used if we had access
+            // to the DS field, and if we set it to one. Indeed,
+            // when DS=1, the shareability fields of page
+            // descriptors are replaced by some bits of the output
+            // address; the shareability is constant for the whole
+            // page table: one for TTBR0 and one for TTBR1.
+            // + TCR_EL1::SH0::Inner
+            // + TCR_EL1::SH1::Inner
+
+            // ASID Size. The upper 8 bits of TTBR0_EL1 and
+            // TTBR1_EL1 are ignored by hardware for every
+            // purpose except reading back the register, and
+            // are treated as if they are all zeros for when
+            // used for allocation and matching entries in the TLB.
+            + TCR_EL1::AS::ASID8Bits
+
+            // We currently output 48 bits of physical
+            // address on aarch64.
+            + TCR_EL1::IPS::Bits_48
+
+            // Translation table walk disable. This bit controls
+            // whether a translation table walk is performed on
+            // a TLB miss.
+            + TCR_EL1::EPD0::EnableTTBR0Walks
+            // + TCR_EL1::EPD1::EnableTTBR1Walks
+
+            // From which TTBR to read the ASID, when comparing the
+            // current address space with the one from an address.
+            + TCR_EL1::A1::TTBR0
+
+            // Controls the size of the memory region addressed
+            // by page table walks. We have to write 64 - (max
+            // number of bits in a region address): 64 - 48 = 16.
+            + TCR_EL1::T0SZ.val(16)
+            // + TCR_EL1::T1SZ.val(16)
+
+            // I (Nathan) don't understand when these flags are used at the moment
+            // + TCR_EL1::ORGN0::WriteBack_ReadAlloc_WriteAlloc_Cacheable
+            // + TCR_EL1::ORGN1::NonCacheable
+            // + TCR_EL1::IRGN0::WriteBack_ReadAlloc_WriteAlloc_Cacheable
+            // + TCR_EL1::IRGN1::NonCacheable
+
+            // Allow the MMU to update the ACCESSED flag.
+            + TCR_EL1::HA::Enable
+
+            // Allow the MMU to update the DIRTY flag.
+            + TCR_EL1::HD::Enable
+        );
+
+        barrier::isb(barrier::SY);
+    }
+}
+
+/// Installs a page table in the TTBR CPU register
+pub fn set_as_active_page_table_root(page_table: PhysicalAddress) {
+    unsafe {
+        let page_table_addr = page_table.value() as u64;
+        TTBR0_EL1.write(
+              TTBR0_EL1::ASID.val(THESEUS_ASID as u64)
+            + TTBR0_EL1::BADDR.val(page_table_addr >> 1)
```

Comment:
Is this correct? I looked up the format of `TTBR0_EL1` and it doesn't seem to indicate that right-shifting it by 1 bit is what we should do.

Reply:
My hint was the authors of `cortex-a` do that [here](https://docs.rs/cortex-a/latest/src/cortex_a/registers/ttbr0_el1.rs.html#60); From the manual, the BADDR field is defined like so:

![image](https://user-images.githubusercontent.com/61582713/208106690-1ec73d3d-9783-4b4e-8bf4-565993da7d86.png)

the least significant bit is omitted; this is explained by the presence of the CnP field (which we don't use) which corresponds to the least significant bit of TTBR0_EL1; that way the MMU doesn't have to shift TTBR0_EL to get BADDR, it's only masked.